packages/firebase_storage/ios/Classes/FirebaseStoragePlugin.m

```diff
@@ -179,21 +194,68 @@ - (void)put:(NSData *)data call:(FlutterMethodCall *)call result:(FlutterResult)
-  [fileRef putData:data
-          metadata:metadata
-        completion:^(FIRStorageMetadata *metadata, NSError *error) {
-          if (error != nil) {
-            result(error.flutterError);
-          } else {
-            [fileRef downloadURLWithCompletion:^(NSURL *URL, NSError *error) {
-              if (error != nil) {
-                result(error.flutterError);
-              } else {
-                result(URL.absoluteString);
-              }
-            }];
-          }
-        }];
+  FIRStorageUploadTask *uploadTask = [fileRef putData:data metadata:metadata];
+  NSNumber *handle = [NSNumber numberWithInt:_nextUploadHandle++];
+  [uploadTask observeStatus:FIRStorageTaskStatusSuccess
+                    handler:^(FIRStorageTaskSnapshot *snapshot) {
+                      [self invokeStorageTaskEvent:handle type:kSuccess snapshot:snapshot];
+                      [self->_uploadTasks removeObjectForKey:handle];
+                    }];
+  [uploadTask observeStatus:FIRStorageTaskStatusProgress
+                    handler:^(FIRStorageTaskSnapshot *snapshot) {
+                      [self invokeStorageTaskEvent:handle type:kProgress snapshot:snapshot];
+                    }];
+  [uploadTask observeStatus:FIRStorageTaskStatusResume
+                    handler:^(FIRStorageTaskSnapshot *snapshot) {
+                      [self invokeStorageTaskEvent:handle type:kResume snapshot:snapshot];
+                    }];
+  [uploadTask observeStatus:FIRStorageTaskStatusPause
+                    handler:^(FIRStorageTaskSnapshot *snapshot) {
+                      [self invokeStorageTaskEvent:handle type:kPause snapshot:snapshot];
+                    }];
+  [uploadTask observeStatus:FIRStorageTaskStatusFailure
+                    handler:^(FIRStorageTaskSnapshot *snapshot) {
+                      [self invokeStorageTaskEvent:handle type:kFailure snapshot:snapshot];
+                      [self->_uploadTasks removeObjectForKey:handle];
+                    }];
+  _uploadTasks[handle] = uploadTask;
+  result(handle);
+}
+
+typedef NS_ENUM(NSUInteger, StorageTaskEventType) {
+  kResume,
+  kProgress,
+  kPause,
+  kSuccess,
+  kFailure
+};
+
+- (void)invokeStorageTaskEvent:(NSNumber *)handle
+                          type:(StorageTaskEventType)type
+                      snapshot:(FIRStorageTaskSnapshot *)snapshot {
+  NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
+  [dictionary setValue:handle forKey:@"handle"];
+  [dictionary setValue:@((int)type) forKey:@"type"];
+  [dictionary setValue:[self buildDictionaryFromTaskSnapshot:snapshot] forKey:@"snapshot"];
+  [self.channel invokeMethod:@"StorageTaskEvent" arguments:dictionary];
+}
+
+- (NSDictionary *)buildDictionaryFromTaskSnapshot:(FIRStorageTaskSnapshot *)snapshot {
+  NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
+  if ([snapshot metadata] != nil && [snapshot.metadata downloadURL] != nil) {
```

Comment:
This is exercising old Firebase API. `FIRStorageMetadata` no longer has a `downloadURL` property. You can run `pod update` to get the latest versions of your pods.